src/domain/covenant-identity.ts

```diff
@@ -0,0 +1,429 @@
+  readonly masterPubKeyNode: Mnemonic['masterPublicKeyNode'];
+  readonly namespace: CovenantIdentityOpts['namespace'];
+  readonly covenant: CovenantDescriptors;
+  readonly ecclib: IdentityOpts<any>['ecclib'] & bip341.TinySecp256k1Interface;
+
+  constructor(args: IdentityOptsWithSchnorr<TemplateIdentityWatchOnlyOpts>) {
+    super(args);
+    this.namespace = args.opts.namespace;
+    this.covenant = {
+      namespace: args.opts.namespace,
+      template: args.opts.template,
+      changeTemplate: args.opts.changeTemplate,
+    };
+    this.ecclib = args.ecclib;
+    this.masterBlindingKeyNode = SLIP77Factory(this.ecclib).fromMasterBlindingKey(
+      args.opts.masterBlindingKey
+    );
+    this.masterPubKeyNode = BIP32Factory(this.ecclib).fromBase58(toXpub(args.opts.masterPublicKey));
+  }
+
+  private deriveMasterXPub(isChange: boolean, index: number, forSchnorr: boolean): string {
+    return this.masterPubKeyNode
+      .derivePath(makePath(isChange, index))
+      .publicKey.slice(forSchnorr ? 1 : 0)
+      .toString('hex');
+  }
+
+  private getContext(isChange: boolean, index: number): Context {
+    return {
+      namespaces: new Map().set(this.namespace, {
+        pubkey: this.deriveMasterXPub(isChange, index, true),
+      }),
+    };
+  }
+
+  protected getBlindingKeyPair(
+    scriptPubKey: string,
+    checkScript = false
+  ): { publicKey: Buffer; privateKey: Buffer } {
+    if (checkScript && !this.cache.has(scriptPubKey))
+      throw new Error(`unknow blinding key for script ${scriptPubKey}`);
+    const { publicKey, privateKey } = this.masterBlindingKeyNode.derive(scriptPubKey);
+    return { publicKey: publicKey!, privateKey: privateKey! };
+  }
+
+  private cacheAddress(addr: TaprootAddressInterface) {
+    const scriptPubKey = address.toOutputScript(addr.confidentialAddress, this.network);
+    this.cache.set(scriptPubKey.toString('hex'), addr);
+  }
+
+  protected getAddressByPublicKey(publicKey: string) {
+    for (const addr of this.cache.values()) {
+      if (addr.publicKey === publicKey) return addr;
+    }
+  }
+
+  private getTemplate(isChange: boolean): string {
+    if (isChange && this.covenant.changeTemplate) return this.covenant.changeTemplate;
+    if (!this.covenant.template) throw new Error('template is missing');
+    return this.covenant.template;
+  }
+
+  getAddress(isChange: boolean, index: number): TaprootAddressInterface {
+    const template = this.getTemplate(isChange);
+    const descriptorResult = evaluate(this.getContext(isChange, index), template);
+    const outputScript = descriptorResult.scriptPubKey().toString('hex');
+    if (!descriptorResult.taprootHashTree) throw new Error('taprootHashTree is missing');
+    return {
+      ...this.outputScriptToAddressInterface(outputScript),
+      result: descriptorResult,
+      derivationPath: makePath(isChange, index),
+      publicKey: this.deriveMasterXPub(isChange, index, true),
```

Comment:
I guess this should be empty for this type of addresses

We should eventually create a dedicated field for the taproot address and we will call `interalPubKey` to be explicit on the meaning of it.